tarantool/connection_pool.py

```diff
@@ -0,0 +1,406 @@
+            if self.RW_iter is not None:
+                return next(self.RW_iter)
+            else:
+                raise ClusterTolopogyError("Can't find healthy rw instance in pool")
+        elif mode == Mode.PREFER_RO:
+            if self.RO_iter is not None:
+                return next(self.RO_iter)
+            elif self.RW_iter is not None:
+                return next(self.RW_iter)
+            else:
+                raise ClusterTolopogyError("Can't find healthy instance in pool")
+        elif mode == Mode.PREFER_RW:
+            if self.RW_iter is not None:
+                return next(self.RW_iter)
+            elif self.RO_iter is not None:
+                return next(self.RO_iter)
+            else:
+                raise ClusterTolopogyError("Can't find healthy instance in pool")
+
+
+@dataclass
+class PoolTask():
+    method_name: str
+    args: tuple
+    kwargs: dict
+
+
+class ConnectionPool(ConnectionInterface):
+    '''
+    Represents pool of connections to the cluster of Tarantool servers.
+
+    ConnectionPool API is the same as a plain Connection API.
+    On each request, a connection is selected to execute this request.
+    Connection is selected based on request mode:
+    * Mode.RW selects an RW instance.
+    * Mode.PREFER_RW selects an RW instance, if possible, RO instance
+      otherwise.
+    * Mode.PREFER_RO selects an RO instance, if possible, RW instance
+      otherwise.
+    All write requests (insert, replace, delete, upsert, update) use
+    RW mode. For any other type of request mode could be set
+    explicitly.
+
+    ConnectionPool updates information about each server state (RO/RW)
+    on initial connect and then asynchronously in separate threads.
+    Application retries must be written considering the asynchronous
+    nature of cluster state refresh. User does not need to use any 
+    synchronization mechanisms in requests, it's all handled with
+    ConnectionPool methods.
+    '''
+    def __init__(self,
+                 addrs,
+                 user=None,
+                 password=None,
+                 socket_timeout=SOCKET_TIMEOUT,
+                 reconnect_max_attempts=POOL_INSTANCE_RECONNECT_MAX_ATTEMPTS,
+                 reconnect_delay=POOL_INSTANCE_RECONNECT_DELAY,
+                 connect_now=True,
+                 encoding=ENCODING_DEFAULT,
+                 call_16=False,
+                 connection_timeout=CONNECTION_TIMEOUT,
+                 strategy_class=RoundRobinStrategy,
+                 refresh_delay=POOL_REFRESH_DELAY):
+        '''
+        Initialize connections to the cluster of servers.
+
+        :param list addrs: List of {host: , port:} dictionaries,
+        describing server addresses.
+        :param int reconnect_max_attempts: Max attempts to reconnect
+        for each connection in the pool. Be careful with reconnect
+        parameters in ConnectionPool since every status refresh is
+        also a request with reconnection. Default is 0.
```

Comment:
Well, at least synchronous `connect` and `close` may stuck for a dead instance. Also some kind of retry logic is already implemented in background loops: we refresh the state each `refresh_delay`. I think that fail-fast on box info fetch in nice here, stucking in retry cycle is not. Moreover, the more retries and retry delays we set, the longer dead instance will be considered alive: at the time of our retries we won't be considering the instance dead (since we haven't got a request response yet) and request may be redirected to its queue.

Reply:
I was in impression that 'default is 0' means 'default is infinite'. Sorry.

Proper timeouts configuration for a connection pool is its own topic. We should come back here later. For now it is okay to just pass existing options to each individual connection.

Reply:
I propose to change 'Default is 0' to 'Default is 0 (fail after first attempt)'.